File: Assignments/01a [Individual] Data parsing servers - Part I/scripts/parse.py
import os
import json
import csv
import yaml
import xml.etree.ElementTree as ET

DATA_FOLDER = r"C:\Users\nicla\OneDrive\Documents\SystemIntegration\SoftwareIntegration\Assignments\01a [Individual] Data parsing servers - Part I\data"

def parse_txt(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        data = f.read().splitlines()
    return data

def parse_json(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        return json.load(f)

def parse_yaml(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        return yaml.safe_load(f)

def parse_csv(file_path):
    with open(file_path, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        return [row for row in reader]

def parse_xml(file_path):
    tree = ET.parse(file_path)
    root = tree.getroot()
    return [elem.text for elem in root.findall("fruit")]

def main():
    files = {
        "txt": parse_txt,
        "json": parse_json,
        "yaml": parse_yaml,
        "csv": parse_csv,
        "xml": parse_xml
    }

    for ext, parser in files.items():
        file_path = os.path.join(DATA_FOLDER, f"data.{ext}")
        try:
            parsed_data = parser(file_path)
            print(f"\nParsed {ext.upper()} file:")
            print(parsed_data)
        except Exception as e:
            print(f"Error parsing {ext.upper()} file: {e}")

if __name__ == "__main__":
    main()

File: Assignments/01a [Individual] Data parsing servers - Part I/scripts/parse.js
const fs = require("fs");
const path = require("path");
const yaml = require("yaml");
const { parseStringPromise } = require("xml2js");

const DATA_FOLDER = path.join(__dirname, "..", "data");

function parseTxt(filePath) {
    return fs.readFileSync(filePath, "utf-8").split("\n").map(line => line.trim()).filter(Boolean);
}

function parseJson(filePath) {
    return JSON.parse(fs.readFileSync(filePath, "utf-8"));
}

function parseYaml(filePath) {
    return yaml.parse(fs.readFileSync(filePath, "utf-8"));
}

function parseCsv(filePath) {
    const data = fs.readFileSync(filePath, "utf-8");
    const lines = data.split("\n").map(line => line.trim()).filter(Boolean);
    const headers = lines.shift().split(",");
    return lines.map(line => {
        const values = line.split(",");
        return headers.reduce((obj, key, i) => ({ ...obj, [key]: values[i] }), {});
    });
}

async function parseXml(filePath) {
    const data = fs.readFileSync(filePath, "utf-8");
    const result = await parseStringPromise(data);
    return result.fruits.fruit;
}

async function main() {
    const files = {
        "txt": parseTxt,
        "json": parseJson,
        "yaml": parseYaml,
        "csv": parseCsv,
        "xml": parseXml
    };

    for (const ext in files) {
        const filePath = path.join(DATA_FOLDER, `data.${ext}`);
        try {
            const parsedData = ext === "xml" ? await files[ext](filePath) : files[ext](filePath);
            console.log(`\nParsed ${ext.toUpperCase()} file:`);
            console.log(parsedData);
        } catch (error) {
            console.error(`Error parsing ${ext.toUpperCase()} file:`, error);
        }
    }
}

main();